fms/models/granite_v4.py

```diff
@@ -0,0 +1,623 @@
+        **kwargs,
+    ):
+        super(GraniteMoeHybrid, self).__init__()
+        if config is not None:
+            self.config = config
+        else:
+            self.config = GraniteMoeHybridConfig()
+        self.config = self.config.updated(**kwargs)
+        self.distributed_strategy = distributed_strategy
+
+        self.base_model = GraniteMoeHybridHeadless(
+            self.config, self.distributed_strategy
+        )
+        self.head = nn.Linear(
+            self.config.emb_dim, self.config.src_vocab_size, bias=False
+        )
+
+    @classmethod
+    def from_config(cls, config: GraniteMoeHybridConfig) -> "GraniteMoeHybrid":
+        return cls(config)
+
+    def get_config(self) -> GraniteMoeHybridConfig:
+        return self.config
+
+    def reset_parameters(self):
+        self.head.weight.data.normal_(
+            0,
+            1 / math.sqrt(math.sqrt(self.config.emb_dim * self.config.src_vocab_size)),
+        )
+        self.base_model.reset_parameters()
+
+    def post_init(self):
+        # if this model ties weights, they are tied here
+        if self.config.tie_heads:
+            # handle assignment of non-meta weights to meta parameters
+            if self.head.weight.device == torch.device("meta"):
+                self.head.weight = self.base_model.embedding.weight
+            else:
+                self.base_model.embedding.weight = self.head.weight
+
+        self.base_model.post_init()
+
+    def forward(
+        self,
+        x: torch.LongTensor,
+        position_ids: Optional[torch.LongTensor] = None,
+        past_key_value_states: Optional[Tuple[torch.FloatTensor,]] = None,
+        use_cache: bool = False,
+        last_n_tokens: int = 0,
+        **attn_kwargs: Unpack[AttentionKwargs],
+    ):
+        get_attention_type(**attn_kwargs)["validate_attn_kwargs"](
+            input_ids=x,
+            position_ids=position_ids,
+            past_key_value_states=past_key_value_states,
+            **attn_kwargs,
+        )
+
+        output, cache = self.base_model(
+            x,
+            position_ids,
+            past_key_value_states,
+            use_cache,
+            **attn_kwargs,
+        )
+
+        output = gather_outputs(output, last_n_tokens, **attn_kwargs)
+        preds = self.head(output)
+        preds = preds / self.config.logits_scaling
+
+        if use_cache:
+            return preds, cache
+        else:
+            return preds
+
+
+_8b_config = GraniteMoeHybridConfig(
+    src_vocab_size=49155,
+    emb_dim=4096,
+    norm_eps=1e-5,
+    nheads=32,
+    kvheads=8,
+    nlayers=40,
+    hidden_grow_factor=12800 / 4096,
+    max_expected_seq_len=8192,
+    rope_theta=10_000.0,
+    pad_id=0,
+    p_dropout=0.0,  # overwriting config.json
+    tie_heads=True,
+    embedding_multiplier=12.0,
+    logits_scaling=16.0,
+    residual_multiplier=0.22,
+    attention_multiplier=0.0078125,
+)
+
+_3_1_2b_config = GraniteMoeHybridConfig(
```

Comment:
_3_1_2b_config is defined but not registered. 
Register it: `models.register_model(_architecture_name, "3.1-2b", ...)`
or remove it from this PR